adding article to connect and upload files programatically to Blob Storage (#5969)
umbraco/UmbracoDocs: umbraco-cloud/set-up/media/connect-programatically-to-azure-storage.md

Reviewer: This article provides the necessary steps needed to programmatically connect to your Umbraco Cloud Environment's Azure Blob Storage containers, in case you want to persist files programmatically.

---
description: "Connect to Cloud environment's Azure Blob Storage programmatically."
---

# Connect to Azure Storage Explorer programmatically

> This article provides the necessary steps needed to programmatically connect to your Azure Blob Storage containers and upload the needed files, without being publicly available. One will need to have access to the credentials of the Blob Storage to authenticate and find the files created programmatically in the Azure Blob Storage.

Author: provide feedback from Tony
```suggestion
This article provides the steps needed to programmatically connect to your Umbraco Cloud Environment's Azure Blob Storage containers, in case you want to persist files programmatically. One will need to have access to the credentials of the Blob Storage to authenticate and find the files created programmatically in the Azure Blob Storage.
```

---
description: "Connect to Cloud environment's Azure Blob Storage programmatically."
---

# Connect to Azure Storage Explorer programmatically

> This article provides the steps needed to programmatically connect to your Umbraco Cloud Environment's Azure Blob Storage containers, to persist files programmatically. You will need access to the Blob Storage credentials to authenticate and find the files created programmatically in the Azure Blob Storage.

Author: ```suggestion
This article provides the steps needed to programmatically connect to your Umbraco Cloud Environment's Azure Blob Storage containers, to persist files programmatically. You will need access to the Blob Storage credentials to authenticate and find the files created programmatically in the Azure Blob Storage.

To connect to your Umbraco Cloud Environment's Azure Blob Storage containers, follow these steps:

```

> 

Author: to not repeat this line, I added "Follow these steps to get started:"

4. Run the project to complete the installation of the package.

5. Add a new class called `BlobStorageService` which serves as a service that has a method to connect to Blob Storage:

{% code title="BlobStorageService.cs" lineNumbers="true" %}

```csharp
using Azure.Storage.Blobs;

namespace UmbracoProject
{
    public class BlobStorageService
    {
        public BlobContainerClient GetContainerClient(string connectionString, string containerName)
        {
            BlobServiceClient blobServiceClient = new BlobServiceClient(connectionString);
            BlobContainerClient containerClient = blobServiceClient.GetBlobContainerClient(containerName);
            return containerClient;

        }
    }
}
```

{% endcode %}

6. Add a new class called `BlobStorageComposer` to inject the service:

{% code title="BlobStorageComposer.cs" lineNumbers="true" %}

```csharp
using Umbraco.Cms.Core.Composing;

namespace UmbracoProject;

public class BlobStorageComposer : IComposer
{
    public void Compose(IUmbracoBuilder builder)
    {
        builder.Services.AddScoped<BlobStorageService>();

    }

> }

Author: hey Esha, from C# 10 or so, you can add ; after namespace and then you can use the classes separately. The code works the same as well. You can read more about this here: https://github.com/umbraco/UmbracoDocs/issues/5465

        BlobContainerClient containerClient = _blobStorageService.GetContainerClient(connectionString, containerName);

        string localPath = "data";
        Directory.CreateDirectory(localPath);
        string fileName = Guid.NewGuid().ToString() + ".txt";
        string localFilePath = Path.Combine(localPath, fileName);

        try
        {
            // Write some content to the file
            await using (StreamWriter writer = new StreamWriter(localFilePath))
            {
                await writer.WriteLineAsync("Hello, World! This file is created programatically!");
            }

        }
        catch (Exception)
        {
            // ignored
        }

        // Get a reference to a blob
        string blobName = "FolderProgramatically/" + Guid.NewGuid().ToString() + ".txt"; //the blobName can be anything
        BlobClient blobClient = containerClient.GetBlobClient(blobName);

        Console.WriteLine("Uploading to Blob storage as blob:\n\t {0}\n", blobClient.Uri);

        // Upload data from the local file
        await blobClient.UploadAsync(localFilePath, true);

        return Content("Check your Blob Storage to see your new file!");
    }

> }

Author: same as previous comment, we should adopt to best practices 👍

{% hint style="warning" %}

From the provided code, replace the content of `string SASUrl` and `string containerName` with the ones from your Umbraco Cloud Settings. You can find more information where you can find those values in the [Connect to Azure Storage Explorer to upload files manually](connect-to-azure-storage-explorer.md#getting-the-credentials) article. 

You can also secure the values in the **secrets setting** on the **project settings** on Umbraco Cloud so you do not store them in code. You can read more about this in the [Secrets Management](../project-settings/secrets-management.md) article.

{% endhint %}

8. Run the project.

9. Visit the `{{yourProjectURL}}/umbraco/surface/BlobStorage/BlobUpdate` endpoint in the backoffice of your project to manually trigger the creation of the file to the Blob Storage.

10. [Connect to your Blob Storage](connect-to-azure-storage-explorer.md) and there you will find the folder and file that has been created programmatically:

![Blob folder created programmatically](images/blob-folder-created-programatically.png)

# References

> The above guide has been created based on the following articles from Microsoft:

Reviewer 2: ```suggestion
For more information, see the following articles from Microsoft:
```
